File: day5.py
from collections import deque


def part1():
    with open('inputs/day5.txt') as f:
        stacks = []

        # There are always just 9 stacks
        for i in range(9):
            stacks.append(deque())

        finished_reading_stacks = False
        for line in f.readlines():
            if line == '\n':
                finished_reading_stacks = True
                continue

            if not finished_reading_stacks:
                if line[0] == '1':
                    # This is the line for the numbered stacks
                    continue
                for i in range(0, len(line) - 1, 4):
                    if line[i] == ' ':
                        # no crate at this position
                        continue
                    stacks[i // 4].appendleft(line[i + 1])
            else:
                # These are the movement procedure lines
                num_to_move = int(line[5:line.index(' ', 5)])
                from_index = line.index('from')
                origin_stack = int(line[from_index + 5])
                dest_stack = int(line[from_index + 10])

                for i in range(num_to_move):
                    stacks[dest_stack - 1].append(stacks[origin_stack - 1].pop())

        print(''.join([stack[-1] for stack in stacks]))


def part2():
    with open('inputs/day5.txt') as f:
        stacks = []

        # There are always just 9 stacks
        for i in range(9):
            stacks.append(deque())

        finished_reading_stacks = False
        for line in f.readlines():
            if line == '\n':
                finished_reading_stacks = True
                continue

            if not finished_reading_stacks:
                if line[0] == '1':
                    # This is the line for the numbered stacks
                    continue
                for i in range(0, len(line) - 1, 4):
                    if line[i] == ' ':
                        # no crate at this position
                        continue
                    stacks[i // 4].appendleft(line[i + 1])
            else:
                # These are the movement procedure lines
                num_to_move = int(line[5:line.index(' ', 5)])
                from_index = line.index('from')
                origin_stack = int(line[from_index + 5])
                dest_stack = int(line[from_index + 10])

                temp = []
                for i in range(num_to_move):
                    temp.append(stacks[origin_stack - 1].pop())

                while len(temp) > 0:
                    stacks[dest_stack - 1].append(temp.pop())

        print(''.join([stack[-1] for stack in stacks]))


if __name__ == "__main__":
    # part1()
    part2()

File: go/day5.go
package main

import (
	"fmt"
	"goaoc22/utils"
	"strconv"
	"strings"
)

func Day5() {
	lines := utils.GetLines("../inputs/Day5.txt")

	finishedReadingStacks := false
	stacks := [9][]string{}
	// for idx := range stacks {
	// 	stacks[idx] = []string{}
	// }

	for _, line := range lines {
		if line == "" {
			continue
		}

		if line[1] == '1' {
			// this is the numbering line
			finishedReadingStacks = true
			continue
		}

		if !finishedReadingStacks {
			for i := 0; i < len(line); i += 4 {
				if line[i] == ' ' {
					// no crate here
					continue
				}

				stacks[i / 4] = append([]string{string(line[i + 1])}, stacks[i / 4]...)
			}
		} else {
			numToMove, err := strconv.Atoi(line[5:utils.IndexAfter(line, " ", 5)])
			if err != nil {
				fmt.Println(err)
				return
			}

			fromStackIdx := int(line[strings.Index(line, "from") + 5] - '0') - 1
			toStackIdx := int(line[strings.Index(line, "from") + 10] - '0') - 1

			for i := 0; i < numToMove; i++ {
				top := stacks[fromStackIdx][len(stacks[fromStackIdx]) - 1]
				stacks[fromStackIdx] = stacks[fromStackIdx][:len(stacks[fromStackIdx]) - 1]
				// fmt.Println("To before: ", stacks[toStackIdx])
				stacks[toStackIdx] = append(stacks[toStackIdx], top)
			}
		}
	}

	var sb strings.Builder
	for _, stack := range stacks {
		sb.WriteString(stack[len(stack) - 1])
	}

	fmt.Println(sb.String())
}